src/boxes/mp4_box.rs

```diff
@@ -0,0 +1,127 @@
+/*
+ * © 2022 Arastoo Bozorgi
+ * All rights reserved.
+ */
+
+use crate::Error;
+use downcast_rs::Downcast;
+
+pub trait Mp4Box: Downcast {
+    /// Read the atom from the data and parse it
+    fn parse(data: &[u8], start: usize, level: u8) -> Result<Self, Error>
+    where
+        Self: Sized;
+
+    /// The start address of the box
+    fn start(&self) -> usize;
+
+    /// The end address of the box
+    fn end(&self) -> usize;
+
+    /// The box size in bytes
+    fn size(&self) -> usize;
+
+    /// The box name
+    fn name(&self) -> &str;
+
+    /// Read the box content
+    fn read(&self) -> Result<Vec<u8>, Error>;
+
+    /// Get the internal boxes of this box
+    fn fields(&self) -> Option<Vec<&Box<dyn Mp4Box>>>;
```

Comment:
warning: you seem to be trying to use `&Box<T>`. Consider using just `&T`
  --> src\boxes\mp4_box.rs:31:36
   |
31 |     fn fields(&self) -> Option<Vec<&Box<dyn Mp4Box>>>;
   |                                    ^^^^^^^^^^^^^^^^ help: try: `&dyn Mp4Box`
   |
   = note: `#[warn(clippy::borrowed_box)]` on by default
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#borrowed_box